posts/Stef/2019-11-18-hibernate-orm-with-panache-in-quarkus.adoc

```diff
@@ -0,0 +1,462 @@
+
+Now, our REST endpoint is not changed much, but let's include it for good measure:
+
+[source%nowrap,java]
+----
+@Path("/")
+public class PersonEndpoint {
+    @Inject
+    private PersonDao personDao;
+    
+    @GET
+    @Path("people")
+    public List<Person> all(){
+        return personDao.findAll().list();
+    }
+
+    @GET
+    @Path("people/by-name")
+    public List<Person> findByName(@PathParam String name){
+        return personDao.findByName(name);
+    }
+
+    @GET
+    @Path("people/born-after")
+    public List<Person> findBornAfter(@PathParam Date date){
+        return personDao.findBornAfter(date);
+    }
+
+    @GET
+    @Path("person/{id}")
+    public Person findById(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        return p;
+    }
+
+    @PUT
+    @Path("person/{id}")
+    public void updatePerson(@PathParam Long id, Person newPerson){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        p.birth = newPerson.birth;
+        p.firstName = newPerson.firstName;
+        p.lastName = newPerson.lastName;
+    }
+
+    @DELETE
+    @Path("person/{id}")
+    public void deletePerson(@PathParam Long id){
+        Person p = personDao.findById(id);
+        if(p == null)
+            throw new WebApplicationException(Status.NOT_FOUND);
+        personDao.delete(p);
+    }
+    
+    @POST
+    @Path("people")
+    public Response newPerson(@Context UriInfo uriInfo, Person newPerson){
+        Person p = new Person();
+        p.birth = newPerson.birth;
+        p.firstName = newPerson.firstName;
+        p.lastName = newPerson.lastName;
+        personDao.persist(p);
+        
+        URI uri = uriInfo.getAbsolutePathBuilder()
+                .path(PersonEndpoint.class)
+                .path(PersonEndpoint.class, "findById")
+                .build(p.id);
+        return Response.created(uri).build();
+    }
+}
+----
+
+This is already a long way towards reducing boilerplate, don't you think?
+
+== Going the extra mile and getting rid of the DAO
+
+Data Access Objects are mostly useful when you have one or more of the following situations:
+
+- The entity type is shared between projects written for different stacks. One project will use DAOs written for WildFly, another for Spring.
+- The entity type is shared between projects written for different use-cases. One project will handle the entity in one way, while another will differ entirely.
+- You need to mock your DAOs in tests.
+- Your entity type is crammed so full of getters and setters that adding any model method will exceed the maximum method count.
+
+While the last reason is a joke, the first reasons are entirely reasonable.
+
+If you are not in one of those situations, DAOs come with drawbacks:
+
+- You need to have one extra class per entity.
```

Comment:
So these one has a bit of an attack surface IMO, as in DDD terms you'd rather have a DAO per aggregate root than per entity. Not sure whether it's a concern for this post but I thought I'd raise it. Btw. it's also why I'm not fully convinced of the "active record" style in Panache. Or is there a way I can have methods relating to contained (sub-)entities also at the aggregate root in Panache?

Reply:
I'm not sure what you mean by aggregate root?

Reply:
I think Gunnar means this:

```
@Entity
public Order extends PanacheEntity {

    @OneToMany(mappedBy = "order")
    public List<OrderLineItem> lineItems;

}
```

```
@Entity 
public OrderLineItem extends PanacheEntity {

   public String label;
   public int amount;
   public BigDecimal price;

}
```

`Order` is the "aggregate root" here. We would usually only have one DAO for both entity types, since line items are aggregated in the order and share its lifecycle.

The point that Panache allows you to write less classes still stands, I think. But as Gunnar said... with Panache, would you be forced to split your queries between `Order` on one side and `OrderLineItem` on the other side? Or can you stuff everything in `Order` like we used to with DAOs (which sort of makes sense, to a point)?

Reply:
you would not be forced to do two DAOs,, you would just write longer queries

```
@Entity
class Order extends PanacheEntity {
    [ ... ] //normal Order fields and queries
    List<OrderLineItem> findOrderLinesNotShipped() {
        return list("select oli from OrderLineItem oli where oli.order_id = ?1 and oli.status = ?2", this.id, Status.NOT_SHIPPED);
    }
}
```

Reply:
> Order is the "aggregate root" here...

Yes, exactly. You don't want to expose entry points for OrderLineItem other than Order or its DAO.

> you would not be forced to do two DAOs...

Ok, interesting. I reckon you'd still have the predefined inherited methods on `OrderLineItem` which might be a downside to DDD purists, but ok. I'd then simply reword the original sentence, as you don't need to have a DAO per entity in the classic style either.